Brain/Brain.ino

```diff
@@ -119,49 +131,32 @@ void loop() {
     packet = make_packet(sensor.id);
     Serial.println(packet);
     RFSerial.println(packet);
-    //bool did_write = write_to_SD(packet);
+    if (!write_to_SD(packet.c_str())){
+      // send some error packet
+    }
   }
 }
 
 
-String file_name = "E1_" + String(month()) + "/" + String(day()) + "/" + String(year()) + "_" + String(hour()) + ":" + String(minute()) + ":" + String(second()) + "_log.txt";
 
-char packet0[] PROGMEM = "Packet0";
-char packet1[] PROGMEM = "Packet1";
-char packet2[] PROGMEM = "Packet2";
-char packet3[] PROGMEM = "Packet3";
-char packet4[] PROGMEM = "Packet4";
-char packet5[] PROGMEM = "Packet5";
-char packet6[] PROGMEM = "Packet6";
-char packet7[] PROGMEM = "Packet7";
-char packet8[] PROGMEM = "Packet8";
-char packet9[] PROGMEM = "Packet9";
 
-int bfr_idx = 0;
-char buffer[64];
-char* packet_table[] PROGMEM = {packet0, packet1, packet2, packet3, packet4, packet5, packet6, packet7, packet8, packet9};
+bool write_to_SD(std::string message) {
+  // every reading that we get from sensors should be written to sd and saved.
 
-bool write_to_SD(String message) {
-  if (!SD.begin(BUILTIN_SDCARD))
-    return false;
-
-  packet_table[bfr_idx] = message.c_str();
-  bfr_idx++;
-
-  if (bfr_idx == 10) {
-    File myFile = SD.open(file_name.c_str(), FILE_WRITE);
-
-    if (myFile) {                                                     //If the file opened
-      for (int i = 0; i <= bfr_idx; i++) {
-        strcpy_P(buffer, (char *)pgm_read_word(&(packet_table[i])));  // Necessary casts and dereferencing, just copy.
-        myFile.println(buffer);
+    sdBuffer->enqueue(message);
+    if(sdBuffer->length >= 40) {
```

Comment:
Is there a reason why >= 40. Is this the similar logic to before where you only write after 10 packets have been written onto?

Reply:
yeah. if we have 40 packets in the queue (in memory), we open the sd and write all 40 at once. This is because opening and closing the sd is pretty slow. 40 is kind of arbitrary; we're optimizing keeping the queue as small as possible (so that we don't lose data in case of crash) and super fast writes which means very infrequent sd open/close, which means very large queue; as you can see they're inversely related.